### ISSUE TITLE
Handle kill signals in the core

### ISSUE THREAD

**Issue by akshgpt7** [metacall/core#121]
### Is your feature request related to a problem? 
If the user presses Ctrl-C in between the execution of an async function inside metacall, the `metacall_destroy()` process doesn't happen, and there's an abrupt exit. This causes many deadlocks and issues, for example [this](https://github.com/metacall/express-faas-rpc-example/issues/2).

### Describe the solution you'd like

The solution should be implemented in the core itself to intercept the kill signal and call `metacall_destroy()` destroy the metacall process gracefully.
The signal handler can be implemented in C in the core (possibly bound to `metacall_initialize()` itself), so that it starts listening as soon as a metacall instance is created. For cross-platform handling, we can refer [this](https://github.com/evgenykislov/ctrl-c).
This can be done in a new `signals` module. For now, we need to intercept `SIGINT` and `SIGTERM` (and maybe `SIGHUP`).

First the loader will handle the Ctrl-C inside itself, then the core will destroy metacall and exit.

### Describe alternatives you've considered
We can think of implementing it just at the loader level (for example, just for Node), but that will be specific to that language, and not very extensible. Implementing it in the core will make it a generalized solution and help later when we have async support for other languages too. Also, we would have to flexibility to handle more signals in future.

**Comment by zikpefu** [comment]
Hi I would like to formally take this issue please, thanks

**Comment by fosslover69** [comment]
Same issue is there with metacall Rust crate too

**Comment by himanshuc3** [comment]
If this isn't resolved, let me dig it up @viferga

**Comment by zikpefu** [comment]
Hey, yeah I completely forgot about this task. Please reassign it from me
please.

**Comment by viferga** [comment]
Not resolved, but we should discuss maybe how to implement it? @zikpefu @himanshuc3 

We should handle this in a cross-platform way in the Core but maybe we have to propagate the signals to node, python etc in order to notify the runtimes, so if you have a signal handler in node, python etc, it works.

Those are the main restrictions, we should think about it and how to do it, then create tests where we test the propagation of the signal into node, python etc.

**Comment by lucasace** [comment]
> Not resolved, but we should discuss maybe how to implement it? @zikpefu @himanshuc3
> 
> We should handle this in a cross-platform way in the Core but maybe we have to propagate the signals to node, python etc in order to notify the runtimes, so if you have a signal handler in node, python etc, it works.
> 
> Those are the main restrictions, we should think about it and how to do it, then create tests where we test the propagation of the signal into node, python etc.

If we implement it in the metacall initialze itself , it would automatically propagate to the other ports right?
something similar had been done with the old backtrace module as well(before it was changed to a plugin extension)

@viferga

**Comment by viferga** [comment]
@lucasace I am not sure 100% we should review it because maybe we need to propagate the signals to the runtimes, but indeed implementing it in the core should solve most of issues.

**Comment by saikat709** [comment]
Hi @zikpefu, are you still working on this issue?

**Comment by zikpefu** [comment]
Hello @saikat709 I am not working on this

**Comment by saikat709** [comment]
Hi @viferga..
What about implementing this with a new "source/signal" module that intercepts "SIGINT"/"SIGTERM", only records the shutdown request in the signal handler, and then performs "metacall_destroy()" from a safe execution context rather than directly inside the handler? This should also let us handle the cross-platform/runtime propagation cleanly. Does this approach make sense?

**Comment by viferga** [comment]
> Hi [@viferga](https://github.com/viferga).. What about implementing this with a new "source/signal" module that intercepts "SIGINT"/"SIGTERM", only records the shutdown request in the signal handler, and then performs "metacall_destroy()" from a safe execution context rather than directly inside the handler? This should also let us handle the cross-platform/runtime propagation cleanly. Does this approach make sense?

There's a problem, when a runtime registers its signal handlers, if we also register them, the order of delivery is unspecified in POSIX. What happens if node signals are triggered before metacall, or the opposite?

And if we have more than one runtime.. it's even more problematic, no?

**Comment by saikat709** [comment]
Yeah. In that case, having both the core and individual runtimes independently register handlers could indeed lead to unpredictable behavior, especially with multiple runtimes.

What about having the core be the single signal handler and then explicitly propagating the received signal to the runtimes that need it, rather than letting each runtime register its own handler? That would give us a deterministic order and avoid multiple independent handlers. Would that be a better direction?

@viferga

**Comment by viferga** [comment]
The way of thinking is this, let's investigate how runtimes work, let's find an unified path that can let us implement it for all and then explode that for reimplementing it for all runtimes.

I do this kind of tricks all around the codebase, it's really unsafe but it pays off. Let's start by NodeJS first:

lib/internal/process.js - setupSignalHandlers()

This is the core JS-level implementation. When you call process.on('SIGTERM', ...), Node intercepts it via a newListener hook and creates a libuv signal wrap:

```js
function setupSignalHandlers() {
  // Load events module in order to access prototype elements on process like
  // process.addListener.
  const signalWraps = {};

  function isSignal(event) {
    return typeof event === 'string' &&
           lazyConstants()[event] !== undefined;
  }

  // Detect presence of a listener for the special signal types
  process.on('newListener', function(type, listener) {
    if (isSignal(type) &&
        !signalWraps.hasOwnProperty(type)) {
      const Signal = process.binding('signal_wrap').Signal;
      const wrap = new Signal();
      wrap.unref();
      wrap.onsignal = function() { process.emit(type); };
      const signum = lazyConstants()[type];
      const err = wrap.start(signum);
      if (err) {
        wrap.close();
        const errnoException = require('util')._errnoException;
        throw errnoException(err, 'uv_signal_start');
      }
      signalWraps[type] = wrap;
    }
  });

  process.on('removeListener', function(type, listener) {
    if (signalWraps.hasOwnProperty(type) &&
        this.listenerCount(type) === 0) {
      signalWraps[type].close();
      delete signalWraps[type];
    }
  });
}   
```

This goes down to libuv which basically is this:
https://docs.libuv.org/en/v1.x/signal.html

It ends up calling to this function:
https://docs.libuv.org/en/v1.x/signal.html#c.uv_signal_start

Here's the function implementation in unix:
https://github.com/libuv/libuv/blob/c2a11ca202e88896669d8bc9fdbb7a62c1ca538b/src/unix/signal.c#L374

And finally, we can see it's using sigaction for registering the signal:
https://github.com/libuv/libuv/blob/c2a11ca202e88896669d8bc9fdbb7a62c1ca538b/src/unix/signal.c#L223

This is the standard in POSIX for signal handling:
https://man7.org/linux/man-pages/man2/sigaction.2.html

Badically all signal handlers for all runtimes will be using this API under the hood in POSIX. If we find a way to intercept this maybe we can hook into the signals of each runtime and override them, in that way we can chose the correct order.

We are already doing this with fork in windows and linux/macos.

But even that I'm not sure if this is still correct.. the ordering must be executed in the same way that the runtimes were initialized maybe and if one application registers exit() in the sigterm it may not execute the subsequent handlers on other runtimes, do you understand? We may need to document that.

**Comment by saikat709** [comment]
Got it. Since the current assignee is no longer working on this, I'm interested in taking it up. Could you assign it to me?

Also, should I start working on the investigation and open a PR once I have a proposed approach, or would you prefer that we discuss the approach here first?

**Comment by viferga** [comment]
> Got it. Since the current assignee is no longer working on this, I'm interested in taking it up. Could you assign it to me?
> 
> Also, should I start working on the investigation and open a PR once I have a proposed approach, or would you prefer that we discuss the approach here first?

I ended up editing my previous post, check it now. No need to assign the issue you can start working on it.

**Comment by viferga** [comment]
The technique for intercepting fork, which we can reuse, it's implemented in windows and linux here:
https://github.com/metacall/core/blob/c397a0f426734515966ea01fd223d011cadd2a9b/source/metacall/source/metacall_fork.c#L272

We could do the same for sigaction. The alternative may be to implement it for each language at standard library level of each runtime.

If you have any other idea I'm open to it.
